### PR TITLE
https://github.com/AdguardTeam/AdguardBrowserExtension/issues/492

### PR DESCRIPTION
**Pull request by atropnikov**

### CODE REVIEW

File: Extension/lib/utils/common.js [hunk 1]
```diff
@@ -337,3 +337,102 @@ StopWatch.prototype = {
         console.log(this.name + "[elapsed: " + elapsed + " ms]");
     }
 };
+
+/**
+ * Utility class for saving and retrieving some details by key;
+ * It's bounded with some capacity.
+ * Details are stored in some ring buffer. For each key corresponding details are retrieved in LIFO order.
+ */
+var DetailsStorage = exports.DetailsStorage = (function () {
```
Comment (ameshkov) [inline Extension/lib/utils/common.js]: Rename to `RingBuffer`

File: Extension/lib/utils/common.js [hunk 2]
```diff
@@ -337,3 +337,102 @@ StopWatch.prototype = {
...
+     * 2. If index has been already in use and details hasn't been processed yet, then removes it from indexes array in detailsKeyToIndex
+     * 3. Push this index to indexes array in detailsKeyToIndex at first position
+     * @param key Key
+     */
+    var create = function (key) {
+
+        var index = detailsWritePointer;
+        detailsWritePointer = (index + 1) % 256;
+
+        var details = detailsRingBuffer[index];
```
Comment (ameshkov) [inline Extension/lib/utils/common.js]: You'd better call it `item` instead of `details`

File: Extension/lib/utils/common.js [hunk 3]
```diff
@@ -337,3 +337,102 @@ StopWatch.prototype = {
...
+        return details;
+    };
+
+    /**
+     * Finds details by key
+     * 1. Get indexes from detailsKeyToIndex by key.
+     * 2. Gets first index from indexes, then gets details from detailsRingBuffer by this index
+     * @param key Key for searching
+     */
+    var get = function (key) {
```
Comment (ameshkov) [inline Extension/lib/utils/common.js]: The thing is that this method removes `key` from index.
So it shouldn't be `get`, method name should be `pop`.

File: Extension/lib/utils/common.js [hunk 4]
```diff
@@ -337,3 +337,102 @@ StopWatch.prototype = {
...
+    }
+
+    /**
+     * Creates and store new details
+     * 1. Associates details with next index from detailsRingBuffer.
+     * 2. If index has been already in use and details hasn't been processed yet, then removes it from indexes array in detailsKeyToIndex
+     * 3. Push this index to indexes array in detailsKeyToIndex at first position
+     * @param key Key
+     */
+    var create = function (key) {
```
Comment (ameshkov) [inline Extension/lib/utils/common.js]: Weird method.

It'd be much easier to understand if it is `put(key, value)`

File: Extension/lib/utils/common.js (hunk 1 above)
Comment (ameshkov) [inline Extension/lib/utils/common.js]: Also we need unit tests for it

File: Extension/browser/firefox/lib/contentPolicy.js [hunk 5]
```diff
@@ -730,26 +731,25 @@ var WebRequestImpl = exports.WebRequestImpl = {
 
         return result;
     },
-    
+
     /**
      * Save request properties to be reused further in http-on-opening-request method.
      * This is ugly, but I have not found a better solution to pass requestType and shouldLoadResult
      * NOTE: Hi, AMO reviewer! Maybe you know a better solution?:)
```
Comment (ameshkov) [inline Extension/browser/firefox/lib/contentPolicy.js]: They don't know, I guess we can remove this comment:)

File: Extension/lib/utils/common.js (hunk 4 above)
Comment (atropnikov) [inline Extension/lib/utils/common.js]: Ok, I'll rename it. But ```value``` is redundant parameter. The main idea is reuse preallocated objects and avoid unnecessary allocations.

Comment (atropnikov) [inline Extension/lib/utils/common.js]: @ameshkov maybe ```push```?

**Pull request merged**